Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,239 @@
+    ///   - logger: The logger.
+    public init(
+        services: [any Service],
+        configuration: ServiceRunnerConfiguration,
+        logger: Logger
+    ) {
+        self.services = services
+        self.configuration = configuration
+        self.logger = logger
+    }
+
+    /// Runs all the services by spinning up a child task per service.
+    /// Furthermore, this method sets up the correct signal handlers
+    /// for graceful shutdown.
+    public func run() async throws {
+        switch self.state {
+        case .initial:
+            self.state = .running
+            try await self._run()
+
+            switch self.state {
+            case .initial, .finished:
+                fatalError("ServiceRunner is in an invalid state \(self.state)")
+
+            case .running:
+                self.state = .finished
+            }
+
+        case .running:
+            throw ServiceRunnerError.alreadyRunning()
+
+        case .finished:
+            throw ServiceRunnerError.alreadyFinished()
+        }
+    }
+
+    private func _run() async throws {
+        self.logger.info(
+            "Starting service lifecycle",
+            metadata: [
+                self.configuration.loggingConfiguration.signalsKey: "\(self.configuration.gracefulShutdownSignals)",
+                self.configuration.loggingConfiguration.servicesKey: "\(self.services)",
+            ]
+        )
+
+        enum ChildTaskResult {
+            case serviceFinished(any Service)
+            case serviceThrew(any Service, any Error)
+            case signalCaught(UnixSignal)
+            case signalSequenceFinished
+        }
+
+        try await withThrowingTaskGroup(of: ChildTaskResult.self) { group in
+            // First we have to register our signals.
+            let unixSignals = await UnixSignalsSequence(trapping: self.configuration.gracefulShutdownSignals)
+
+            group.addTask {
+                for await signal in unixSignals {
+                    return .signalCaught(signal)
+                }
+
+                return .signalSequenceFinished
+            }
+
+            for service in self.services {
+                self.logger.debug(
+                    "Starting service",
+                    metadata: [
+                        self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                    ]
+                )
+
+                group.addTask {
+                    do {
+                        try await service.run()
+                        return .serviceFinished(service)
+                    } catch {
+                        return .serviceThrew(service, error)
+                    }
+                }
+            }
+
+            var finishedServiceCount = 0
+
+            // We are going to wait for any of the services to finish or
+            // the signal sequence to throw an error.
+            while !group.isEmpty {
+                let result = try await group.next()
```

Comment:
I think we need to use `nextResult()!` here. As a throw would leave this scope!

Reply:
This is a bit annoying. No child task is actually throwing here. We should probably just use `try!`